README.md
# THIS BOT IS OUTDATED. NEW VERSION AVAILABLE HERE https://github.com/kyllian1212/Rin-rewrite

# Rin

#### NOTE: this bot is made for the Porter Robinson Discord and the r/Madeon Discord only. the code is public because i thought it'd be interesting to share it, but it will not work on other servers.


changelog.md
# Old Rin changelog

This was a very early version of Rin, created as an emergency after the old Sad Machine bot died, to keep the report features working. It wasn't announced publically at all in either the Porter Robinson Discord or the Madeon Discord until v1.0.0. 

Fun fact: most of the functionality on those old Rin versions was contained in only one file! (which was a mess)

(date format is dd/mm/yyyy)

------------------------------------------------------------------------
## v0.3.20-final - 10/09/2021
* added a countdown for second sky
* this is the final version of the bot (before rewritting) due to discord.py being deprecated
## v0.3.19 - 11/07/2021
* removal of the code for the PRD event announcement "game"
* some changes to the dev mode, to separate it from Rin in the PRD and test stuff elsewhere
* actual new stuff for the bot coming Soon:tm:
## v◈.◈.6 - 10/07/2021
* PRD event announcement "game" part 6
## v◈.◈.5 - 09/07/2021
* PRD event announcement "game" part 5
## v◈.◈.4 - 08/07/2021
* PRD event announcement "game" part 4
## v◈.◈.3 - 07/07/2021
* PRD event announcement "game" part 3
## v◈.◈.2 - 06/07/2021
* PRD event announcement "game" part 2
## v◈.◈.1 - 05/07/2021
* PRD event announcement "game" part 1
## v◈.◈.◈ - 04/07/2021
* PRD event announcement "game" teasing
## v0.3.18 - 02/06/2021
* added the 'sky marshall' role to the list of roles that allows you to delete messages when you report them
## v0.3.17 - 29/05/2021
* removed nurture countdown stuff
* reverted rin's rich presence to random songs instead of nurture in order
## v0.3.16 - 21/04/2021
* updated countdown stuff
* the bot will now play nurture in order in the rich presence
## v0.3.15 - 14/04/2021
* updated the report functionnality so that temp moderators can report messages as well
## v0.3.14 - 07/04/2021
* fixed !!nurture's issue (again x2) with counting seconds/days
## v0.3.13 - 07/04/2021
* added !!saytts (for mods)
* added special countdown stuff for nurture's release
* switched the timezone of the countdown from utc to nzst
## v0.3.12 - 22/03/2021
* renamed the !!days_to_nurture command to !!nurture
* fixed !!nurture's issue (again) with counting seconds
## v0.3.11 - 22/03/2021
* fixed !!days_to_nurture's issue with counting days
## v0.3.10 - 21/03/2021
* updated the bot's functionment (related to dev mode and versionning)
* updated !!days_to_nurture and it's functionment  
* updated !!fiftyfifty's functionment (will draw a number from 0 to 999 instead of 1 or 2, 0 to 499 being yes and 500 to 999 being no) 
* added placeholder stuff
## v0.3.9 - 12/03/2021
* added musician by porter robinson to the song library
* added !!fiftyfifty for geoguessr in #live-games
## v0.3.8 - 28/02/2021
* migrated from heroku to google cloud, which removed the 24hrs reboot limitation. i removed it from the bot as well as a result
## v0.3.7 - 13/02/2021
* fixed a glitch with the 'days to nurture' automation in #nurture
* fixed the song length for 'look at the sky' (5m10s instead of 2m32s)
* added placeholder code for when nurture releases
## v0.3.6 - 11/02/2021
* updated libraries, including discord.py to 1.6.0
* a reported message will now be deleted instead of staying up if a mod reacts 🚫 after a regular user did
* added a countdown to be posted in #nurture every day at 12:00am GMT. closer to the release i'll add a function that will allow the bot to send a message for each timezone that'll get nurture.
* added !!days_to_nurture that will tell you how many days are left before nurture releases (in the GMT timezone)
## v0.3.5 - 15/12/2020
* added a very barebone !!roll command that throws a d6 dice
* the bot now shows the version in it's nickname
* !!lastchange has been removed and it's functionality was moved to !!changelog (i felt like both commands were redundant)
## v0.3.4 - 11/12/2020
* the bot will now automatically restart at a certain undisclosed time
* little functional change to !!lastchange (just so that it's easier for me to edit versions lmfao)
## v0.3.3 - 06/12/2020
* added !!lastchange
* added madeon's adventure (deluxe) into the bot's presence
* changed the bot's presence format ('song' by 'artist')
* fixed an issue where the bot would crash if someone reacted to a webhook/deleted user 
* fixed the changelog links oops
* normalized date format to dd/mm/yyyy
## v0.3.2 - 11/11/2020 
* added a version verification (if versions aren't matching, the bot will quit)
* added !!changelog
* full changelog is now in a markdown format
* fixed the issue with the bot crashing if people react in another server
* fixed the issue with edited messages not being able to be reported
## v0.3.1 - 02/11/2020
* fixed the !!info command (everyone used to have created the bot lmao)
## v0.3.0 - 01/11/2020
* added !!say command
* added !!info command
* added !!october18 command
* transferred the version into an .env variable
* the bot will now show when it's in dev mode (as in when it's running locally, rather than running on heroku)

------------------------------------------------------------------------
## v0.2.1 - 01/11/2020 
* converted the !!quit command into a proper bot.command() instead of a on_message() command
* .gitignore update to exclude the python cache
## v0.2.0 - 01/11/2020 
* added more songs into the library
* added a proper crash handler


------------------------------------------------------------------------
## v0.1.2 - 31/10/2020
* added songs into the presence thing
* bugfix with crash issue 
## v0.1.1 - 31/10/2020
* bugfix with role issues
## v0.1.0 - 31/10/2020
* initial release with report function


song_library.py
'''
how to convert a song's length to minutes:seconds to seconds: 
0-60 if the song lasts anywhere between 0 second and a minute
(60*minutes)+seconds otherwise
'''
def convert(minutes, seconds):
    return (60*minutes)+seconds
    

artist_library=[
    "Porter Robinson",
    "VIRTUAL SELF",
    "Madeon",
    "Porter Robinson & Madeon"
]

song_library=[
        ##porter robinson
        #pre-spitfire
        ["Say My Name", artist_library[0], convert(6,32)],
        #spitfire
        ["Spitfire", artist_library[0], convert(6,45)],
        ["Unison", artist_library[0], convert(5,50)],
        ["100% in the Bitch", artist_library[0], convert(4,11)],
        ["Vandalism (ft. Amba Shepherd)", artist_library[0], convert(7,20)],
        ["The State", artist_library[0], convert(5,57)],
        ["The Seconds (ft. Jano)", artist_library[0], convert(5,43)],
        #language + easy
        ["Language", artist_library[0], convert(6,8)],
        ["Easy (with Mat Zo)", artist_library[0], convert(6,36)],
        #worlds
        ["Divinity (ft. Amy Millan)", artist_library[0], convert(6,8)],      
        ["Sad Machine", artist_library[0], convert(5,50)],                  
        ["Years of War (ft. Breanne Düren and Sean Caskey)", artist_library[0], convert(3,56)],                 
        ["Flicker", artist_library[0], convert(4,39)],                      
        ["Fresh Static Snow", artist_library[0], convert(5,58)],            
        ["Polygon Dust (ft. Lemaitre)", artist_library[0], convert(3,29)],            
        ["Hear The Bells (ft. Imaginary Cities)", artist_library[0], convert(4,46)],            
        ["Natural Light", artist_library[0], convert(2,58)],            
        ["Lionhearted (ft. Urban Cone)", artist_library[0], convert(4,24)],  
        ["Sea of Voices", artist_library[0], convert(4, 59)],          
        ["Fellow Feeling", artist_library[0], convert(5, 49)],    
        ["Goodbye to a World", artist_library[0], convert(5, 28)],   
        #virtual self
        ["Particle Arts", artist_library[1], convert(2,58)],            
        ["Ghost Voices", artist_library[1], convert(4,24)],  
        ["a.i.ngel (Become God)", artist_library[1], convert(4, 59)],           
        ["Key", artist_library[1], convert(5, 28)],  
        ["EON BREAK", artist_library[1], convert(2,58)],         
        ["ANGEL VOICES", artist_library[1], convert(6, 32)],      
        #nurture
        ["Lifelike", artist_library[0], convert(1, 35)],
        ["Look At The Sky", artist_library[0], convert(5, 10)],  
        ["Get Your Wish", artist_library[0], convert(3, 38)],  
        ["Wind Tempos", artist_library[0], convert(6, 4)],
        ["Musician", artist_library[0], convert(3, 58)],
        ["do-re-mi-fa-so-la-ti-do", artist_library[0], convert(3, 35)],
        ["Mother", artist_library[0], convert(3, 36)],
        ["dullscythe", artist_library[0], convert(4, 0)],
        ["Sweet Time", artist_library[0], convert(4, 12)],
        ["Mirror", artist_library[0], convert(5, 7)],
        ["Something Comforting", artist_library[0], convert(4, 42)],  
        ["Blossom", artist_library[0], convert(3, 46)],
        ["Unfold (with Totally Enormous Extinct Dinosaurs)", artist_library[0], convert(4, 46)],
        ["Trying to Feel Alive", artist_library[0], convert(4, 44)],
        ["fullmoon lullaby (with Wednesday Campanella)", artist_library[0], convert(4, 3)],
        ##porter robinson & madeon (2016)
        ["Shelter", artist_library[3], convert(3, 39)],  

        ##madeon
        #adventure
        ["Isometric (Intro)", artist_library[2], convert(1,20)],      
        ["You're On (ft. Kyan)", artist_library[2], convert(3,12)],                  
        ["OK", artist_library[2], convert(3,1)],                 
        ["La Lune (ft. Dan Smith)", artist_library[2], convert(3,39)],                      
        ["Pay No Mind (ft. Passion Pit)", artist_library[2], convert(4,9)],            
        ["Beings", artist_library[2], convert(3,35)],            
        ["Imperium", artist_library[2], convert(3,18)],            
        ["Zephyr", artist_library[2], convert(3,40)],            
        ["Nonsense (ft. Mark Foster)", artist_library[2], convert(3,45)],  
        ["Innocence (ft. Aquilo)", artist_library[2], convert(3, 44)],    
        ["Pixel Empire", artist_library[2], convert(4,4)],      
        ["Home", artist_library[2], convert(3,45)],                  
        ["Icarus", artist_library[2], convert(3,34)],                 
        ["Finale (ft. Nicholas Petricca)", artist_library[2], convert(3,24)],                      
        ["The City", artist_library[2], convert(3,54)],            
        ["Cut the Kid", artist_library[2], convert(3,17)],            
        ["Technicolor", artist_library[2], convert(6,25)],            
        ["Only Way Out (ft. Vancouver Sleep Clinic)", artist_library[2], convert(3,46)],              
        #good faith
        ["Dream Dream Dream", artist_library[2], convert(3,54)],      
        ["All My Friends", artist_library[2], convert(3,24)],                  
        ["Be Fine", artist_library[2], convert(3,28)],                 
        ["Nirvana", artist_library[2], convert(2,32)],                      
        ["Mania", artist_library[2], convert(2,32)],            
        ["Miracle", artist_library[2], convert(4,10)],            
        ["No Fear No More", artist_library[2], convert(3,15)],            
        ["Hold Me Just Because", artist_library[2], convert(3,6)],            
        ["Heavy With Hoping", artist_library[2], convert(4,1)],  
        ["Borealis", artist_library[2], convert(4, 45)],    
    ]

main.py
# main.py
import os

import discord
from dotenv import load_dotenv
from datetime import datetime
from discord.ext import commands
from discord.ext.commands import bot
import sys
import asyncio
import traceback
import random

import song_library as sl

load_dotenv()
TOKEN = os.getenv('DISCORD_TOKEN')
GUILD = os.getenv('DISCORD_GUILD')

#make sure to change the version when updated!
version = "v0.3.20-final"
version_date = "10/09/2021"

#dev mode is when i run the bot (dont forget to disable it!!!)
devmode = False

intents = discord.Intents.all()
bot = commands.Bot(command_prefix="!!", intents=intents)

@bot.event
async def on_ready():
    try:
        global devmode
        global version
        now = datetime.now()

        for guild in bot.guilds:
            if guild.name == GUILD:
                break

        print(
            f'{bot.user} is connected to the following guild:\n'
            f'{guild.name}(id: {guild.id})\n'
        )

        membercount = 0
        for member in guild.members:
            membercount = membercount+1

        print(f'Guild Member Count: {membercount}')

        #version mismatch check (deprecated)
        '''
        if VERSION != GIT_VERSION:
            if devmode:
                versionmismatch_message_embed = discord.Embed(title="warning: versions aren't matching", description="*the bot is in dev mode, so it will boot anyways.*", color=0xffae42)
            else:
                versionmismatch_message_embed = discord.Embed(title="error: versions aren't matching!!", color=0xff0000)
                versionmismatch_message_embed.add_field(name="reason", value="this error can happen because:\n " + 
                "- the .env value for 'version' isn't correctly set\n" +
                "- the bot is running on the previous commit\n " +
                "- the bot is in dev mode, but there's an issue with the devmode variable\n " +
                "\nplease check the .env values or make sure the next commit is built. the bot will now shutdown."
                , inline=False)
            versionmismatch_message_embed.add_field(name=".env version", value="*"+VERSION+"*", inline=True)
            versionmismatch_message_embed.add_field(name="git version", value="*"+GIT_VERSION+"*", inline=True)
            versionmismatch_message_embed.set_footer(text=str(now.strftime("%d/%m/%Y - %H:%M:%S")))

            await bot.get_channel(772219545082396692).send(embed=versionmismatch_message_embed)

            if not devmode:
                await os._exit(-1)
        '''

        #devmode check for versioning indication
        if devmode:
            version = version + "-dev"

        #bot version to be changed here for now - v major.minor.bugfix-dev
        init_message_embed = discord.Embed(title="bot has successfully booted up.", description="*bot version: " + version + "*", color=0x00aeff)
        init_message_embed.set_footer(text=str(now.strftime("%d/%m/%Y - %H:%M:%S")))

        bot.loop.create_task(status_task())

        #full boot sequence successfully completed
        if not devmode:
            await bot.get_channel(772219545082396692).send(embed=init_message_embed)
            await bot.get_guild(186610204023062528).get_member(307932112294772737).edit(nick="Rin | " + version)
        else:
            await bot.get_channel(855228682166599741).send(embed=init_message_embed)
            await bot.get_guild(849034525861740571).get_member(849410467507601459).edit(nick="Rin | " + version)
        
    except:
        await crash_handler()
        await os._exit(-1)
        raise

@bot.command()
async def quit(ctx):
    try:
        if not devmode:
            now = datetime.now()
            guild = bot.get_guild(186610204023062528)
            mod_role = guild.get_role(219977258453041152)
            message_author = ctx.author
            is_mod = False
            for role in message_author.roles:
                if role == mod_role:
                    is_mod = True
            if is_mod:
                await ctx.channel.send("bot has been shutdown.")
                quit_message_embed = discord.Embed(title="bot has successfully shutdown.", color=0x00aeff)
                quit_message_embed.set_footer(text=str(now.strftime("%d/%m/%Y - %H:%M:%S")))
                await bot.get_channel(772219545082396692).send(embed=quit_message_embed)
                await bot.loop.stop()
                await os._exit(1)
            else:
                await mod_only_command(ctx)
        else:
            now = datetime.now()
            await ctx.channel.send("bot has been shutdown.")
            quit_message_embed = discord.Embed(title="bot has successfully shutdown.", color=0x00aeff)
            quit_message_embed.set_footer(text=str(now.strftime("%d/%m/%Y - %H:%M:%S")))
            await bot.get_channel(855228682166599741).send(embed=quit_message_embed)
            await bot.loop.stop()
            await os._exit(1)
    except:
        await crash_handler()
        raise

@bot.command()
async def say(ctx, *, arg):
    try:
        if not devmode:
            guild = bot.get_guild(186610204023062528)
            mod_role = guild.get_role(219977258453041152)
            message_author = ctx.author
            is_mod = False
            for role in message_author.roles:
                if role == mod_role:
                    is_mod = True
            if is_mod:
                await ctx.message.delete()
                await ctx.channel.send(arg)
        else:
            await ctx.message.delete()
            await ctx.channel.send(arg)
    except:
        await crash_handler()
        raise

@bot.command()
async def saytts(ctx, *, arg):
    try:
        if not devmode:
            guild = bot.get_guild(186610204023062528)
            mod_role = guild.get_role(219977258453041152)
            message_author = ctx.author
            is_mod = False
            for role in message_author.roles:
                if role == mod_role:
                    is_mod = True
            if is_mod:
                await ctx.message.delete()
                await ctx.channel.send(arg, tts=True)
        else:
            await ctx.message.delete()
            await ctx.channel.send(arg, tts=True)
    except:
        await crash_handler()
        raise

@bot.command()
async def info(ctx):
    try:
        now = datetime.now()
        kyllian_user = bot.get_user(171000921927581696)
        info_message_embed = discord.Embed(title="Rin-bot by " + str(kyllian_user.name) + "#" + str(kyllian_user.discriminator), description="*bot version " + version + "*", url="https://github.com/kyllian1212/Rin", color=0x00aeff)
        info_message_embed.set_thumbnail(url=kyllian_user.avatar_url)
        info_message_embed.set_footer(text=str(now.strftime("%d/%m/%Y - %H:%M:%S")) + "  •  source code available by clicking the link above", icon_url=bot.user.avatar_url)
        await ctx.channel.send(embed=info_message_embed)
    except:
        await crash_handler()
        raise

@bot.command()
async def october18(ctx):
    try:
        dad_message_embed = discord.Embed(title="To: Rin", description="**From: Dad**", color=0x00aeff)
        dad_message_embed.set_thumbnail(url=bot.user.avatar_url)
        dad_message_embed.add_field(name="Message", value="There was just so little time left after you were born. \nI don't know how much love I managed to pour into raising you after your mother died... \nBut your smile kept me going. (^_^) \n\nI would like to have come with you, but I couldn't. \nI wanted you to forget everything and move on... \nI knew you'd be alright. But you'll get lonely, and remember. \n\nI know you'll grow strong, and read this letter some day. \nI really wish we could have spent more time together. I'm sorry. \nYou were so young back then, too young to understand what they meant. So let me repeat... \n\nMy final words to you.", inline=False)
        dad_message_embed.set_footer(text="19/10/2016 - 04:00:00 (JST)")
        await ctx.channel.send(embed=dad_message_embed)
    except:
        await crash_handler()
        raise

@bot.command()
async def changelog(ctx):
    try:
        now = datetime.now()
        lastversion = version + " - " + version_date
        changelog = open('lastchange_bot.txt', 'r').read()
        changelog_message_embed = discord.Embed(title="hello i've updated the bot :) | " + lastversion, description=changelog, url="https://github.com/kyllian1212/Rin/blob/master/changelog.md", color=0x00aeff)
        changelog_message_embed.set_thumbnail(url=bot.user.avatar_url)
        changelog_message_embed.set_footer(text=str(now.strftime("%d/%m/%Y - %H:%M:%S")) + "  •  full changelog available by clicking the link above")
        await ctx.channel.send(embed=changelog_message_embed)
    except:
        await crash_handler()
        raise

@bot.command()
async def roll(ctx):
    try:
        random_variable = random.randint(1, 6)
        await ctx.channel.send("1d6 roll result: **" + str(random_variable) + "**")
    except:
        await crash_handler()
        raise

@bot.command()
async def fiftyfifty(ctx):
    try:
        random_variable = random.randint(0, 999)
        if 0 <= random_variable <= 499:
            await ctx.channel.send("50/50 choice: **YES**")
        if 500 <= random_variable <= 999:
            await ctx.channel.send("50/50 choice: **NO**")
    except:
        await crash_handler()
        raise

@bot.event
async def on_reaction_add(reaction, user):
    try:
        now = datetime.now()
        reacted_message = reaction.message
        reacted_message_content = reacted_message.content
        reacted_message_author = reaction.message.author
        reacted_message_author_roles = None
        guild = bot.get_guild(186610204023062528)
        mod_role = guild.get_role(219977258453041152)
        temp_mod_role = guild.get_role(816596360064532501)
        sky_marshal = guild.get_role(782224911648948244)

        if reaction.count == 1:
            if devmode:
                channel_report = bot.get_channel(855228720154935316)
            else:
                channel_report = bot.get_channel(413488194819194891)
            long_message = False
            mod_report = False
            description = ""

            #checks if the user is a webhook/deleted and doesnt report if so
            if reacted_message_author.discriminator != '0000':
                if user.guild == guild:
                    for role in reacted_message_author.roles:
                        if role.name != '@everyone':
                            if reacted_message_author_roles is None:
                                reacted_message_author_roles = str(role)
                            else:
                                reacted_message_author_roles = str(reacted_message_author_roles) + ", " + str(role)

                    
                    if reaction.emoji == '🚫':
                        #if a user auto reacts 🚫 to their message, it just deletes it without reporting it
                        if user == reacted_message_author:
                            await reacted_message.delete()
                        else:
                            for role in user.roles:
                                if role == mod_role or role == temp_mod_role or role == sky_marshal:
                                    #report_confirmation_embed = discord.Embed(title="The message has been reported.", color=0xff0000)
                                    await reacted_message.delete()
                                    mod_report = True
                                    description = "*The message has been deleted as it has been reported by an admin.*"
                                    #report_confirmation_message = await reacted_message.channel.send(embed=report_confirmation_embed)
                                    #await asyncio.sleep(2)
                                    #await report_confirmation_message.delete()

                            reported_message_embed = discord.Embed(title="Reported Message - ID: " + str(reacted_message.id), description=description, color=0xff0000)
                            reported_message_embed.set_thumbnail(url=reacted_message_author.avatar_url)
                            reported_message_embed.add_field(name="Username", value=reacted_message_author.name, inline=True)
                            reported_message_embed.add_field(name="Nickname", value=reacted_message_author.nick, inline=True)
                            reported_message_embed.add_field(name="User ID", value="<@" + str(reacted_message_author.id) + ">", inline=True)
                            reported_message_embed.add_field(name="Channel", value="<#" + str(reacted_message.channel.id) + ">", inline=True)
                            reported_message_embed.add_field(name="User Roles", value=str(reacted_message_author_roles), inline=True)
                            #if the message is too long
                            if len(reacted_message_content) >= 1000:
                                long_message = True
                                message_length = len(reacted_message_content)
                                reacted_message_content_part2 = reacted_message_content[999:message_length]
                                reacted_message_content = reacted_message_content[0:999]
                                reported_message_embed.add_field(name="Message", value=reacted_message_content, inline=False)
                                reported_message_part2_embed = discord.Embed(description="*The message is over 1000 characters, so it has been split into 2 Discord embeds.*", color=0xff0000)
                                reported_message_part2_embed.add_field(name="Message (part 2)", value=reacted_message_content_part2, inline=False)
                                reported_message_part2_embed.set_footer(text="Reported by " + str(user.name) + "#" + str(user.discriminator) + "  •  " + str(now.strftime("%d/%m/%Y - %H:%M:%S")), icon_url=user.avatar_url)
                            else:
                                #if the message only contains an attachment
                                if len(reacted_message_content) != 0:
                                    reported_message_embed.add_field(name="Message", value=reacted_message_content, inline=False)
                                else:
                                    reported_message_embed.add_field(name="Attachment", value="*The reported message only contained an attachment; make sure to have it saved before reporting it as the bot cannot currently save (deleted) pictures. You can also give out a description of the attachment below*", inline=False)
                                reported_message_embed.set_footer(text="Reported by " + str(user.name) + "#" + str(user.discriminator) + "  •  " + str(now.strftime("%d/%m/%Y - %H:%M:%S")), icon_url=user.avatar_url)

                            await channel_report.send(embed=reported_message_embed)
                            if long_message == True:
                                await channel_report.send(embed=reported_message_part2_embed)
        elif reaction.count > 1:
            if reaction.emoji == '🚫':
                #if a user auto reacts 🚫 to their message, it just deletes it without reporting it
                if user == reacted_message_author:
                    await reacted_message.delete()
                else:
                    for role in user.roles:
                        if role == mod_role or role == temp_mod_role or role == sky_marshal:
                            await reacted_message.delete()

    #put crash handler in another function   
    except:
        await crash_handler()
        raise

@bot.event
async def mod_only_command(message):
    await message.channel.send("you cannot do this action as you are not a mod!")

#rin presence
async def status_task():
    try:
        await bot.wait_until_ready()
        while True:
            variable = random.randint(0, (len(sl.song_library)-1))
            if str(version).endswith("-dev"):
                await bot.change_presence(activity=discord.Activity(type=discord.ActivityType.listening, name="[DEV MODE] " + sl.song_library[variable][0] + " by " + sl.song_library[variable][1]))
            else:
                await bot.change_presence(activity=discord.Activity(type=discord.ActivityType.listening, name=sl.song_library[variable][0] + " by " + sl.song_library[variable][1]))
            #DONT FORGET TO AWAIT A ASYNCIO.SLEEP() COMMAND!!!!!
            await asyncio.sleep(sl.song_library[variable][2])
    except:
        await crash_handler()
        await status_task()
        raise

#crash handler
async def crash_handler():
    try:
        now = datetime.now()
        crash_traceback = traceback.format_exc()
        crash_message_embed = discord.Embed(title="it crashed :(", description=crash_traceback, color=0xff0000)
        crash_message_embed.set_footer(text=str(now.strftime("%d/%m/%Y - %H:%M:%S")))
        if not devmode: await bot.get_channel(772219545082396692).send(embed=crash_message_embed)
        else: await bot.get_channel(855228682166599741).send(embed=crash_message_embed)
    except:
        try:
            now = datetime.now()
            crash_traceback = traceback.format_exc()
            crash_message_embed = discord.Embed(title="the crash handler crashed!! D:", description=crash_traceback, color=0xff0000)
            crash_message_embed.set_footer(text=str(now.strftime("%d/%m/%Y - %H:%M:%S")))
            if not devmode: await bot.get_channel(772219545082396692).send(embed=crash_message_embed)
            else: await bot.get_channel(855228682166599741).send(embed=crash_message_embed)
        except:
            raise

#launch bot
if __name__ == "__main__":
    try:
        bot.run(TOKEN)
    except:
        raise

#bot log: tmux a -t [number]
